pypicloud/access/aws_secrets_manager.py

```diff
@@ -0,0 +1,181 @@
+              "user5": ["read"],
+          }
+      }
+  }
+}
+"""
+
+
+class AWSSecretsManagerAccessBackend(IAccessBackend):
+
+    """
+    This backend allows you to store all user and package permissions in
+    AWS Secret Manager
+
+    """
+
+    def __init__(self, request=None, settings=None, region=None,
+                 secret_id=None, credentials=None, **kwargs):
+        super(AWSSecretsManagerAccessBackend, self).__init__(request, **kwargs)
+        self._settings = settings
+        self.region = region
+        self.secret_id = secret_id
+        self.credentials = {}
+        self.credentials.update(credentials or {})
+
+    @classmethod
+    def configure(cls, settings):
+        kwargs = super(AWSSecretsManagerAccessBackend, cls).configure(settings)
+        kwargs['settings'] = settings
+        kwargs['region'] = settings['auth.region']
+        kwargs['secret_id'] = settings['auth.secret_id']
+        credentials = {}
+
+        if settings.get('auth.access_key'):
+            credentials['aws_access_key_id'] = settings['auth.access_key']
+        if settings.get('auth.secret_key'):
+            credentials['aws_secret_access_key'] = settings['auth.secret_key']
+        if settings.get('auth.session_token'):
+            credentials['aws_session_token'] = settings['auth.session_token']
+        if settings.get('auth.profile_name'):
+            credentials['profile_name'] = settings['auth.profile_name']
+        kwargs['credentials'] = credentials
+        return kwargs
+
+    def _fetch_credentials(self):
+        """ Hit a server endpoint and return the json response """
+        session = boto3.session.Session(**self.credentials)
+        client = session.client(
+            service_name='secretsmanager',
+            region_name=self.region,
+            endpoint_url="https://secretsmanager.{}.amazonaws.com".format(
+                self.region
+            )
+        )
+
+        try:
+            return json.loads(client.get_secret_value(
+                SecretId=self.secret_id
+            )['SecretString'])
+        except ClientError as e:
+            if e.response['Error']['Code'] == 'ResourceNotFoundException':
+                raise Exception(
+                    "The requested secret " + self.secret_id +
+                    " was not found")
+            elif e.response['Error']['Code'] == 'InvalidRequestException':
+                raise Exception("The request was invalid due to:", e)
+            elif e.response['Error']['Code'] == 'InvalidParameterException':
+                raise Exception("The request had invalid params:", e)
+
+    def verify_user(self, username, password):
+        credentials = self._fetch_credentials()
+        try:
+            return credentials.get('users', {})[username] == password
+        except KeyError:
+            return False
+
+    def _get_password_hash(self, username):
+        # We don't have to do anything here because we overrode 'verify_user'
+        pass
```

Comment:
As @tomelliff mentioned, it makes more sense to store the passwords hashed. It should be the same as the instructions for using the [config file for auth](http://pypicloud.readthedocs.io/en/latest/topics/access_control.html#user-username). You run `ppc-gen-password` for each of the users instead of putting the raw password into the json blob.

Some background on passlib: the default format that we're using here *does* include a salt, so that's working as intended. The [default number of rounds](https://github.com/stevearc/pypicloud/blob/e9b1d3d284a65f9ad8767b64044c2b1bbeecc315/pypicloud/access/base.py#L17) that we've chosen for pypicloud is based on [the default rounds provided by passlib](https://bitbucket.org/ecollins/passlib/src/4801587cebf01f5037ddc9cd52fc94708559bbfb/passlib/apps.py?at=stable&fileviewer=file-view-default#apps.py-71:85), but was made customizable for #115.

Passlib can choose *secure* defaults for the rounds, but it wouldn't be able to choose *smart* defaults because the rounds are determined at the time of creating the password hash. So in order to do anything smart like choosing a number of rounds that corresponds to a specific number of milliseconds, you would have to be generating the hashes on the same machine that is going to be hosting your webserver. For this and the config file backend, I do not expect that will often be the case.

Reply:
I agree, and actually found methods to create and update passwords for a user in the base class (that I didn't find before as I started from the external api class), but I couldn't found the routes or the interface to use them. How do you do that?

Also, if we're not using the password rotation feature from AWS secret manager, I'm thinking of moving the secrets to SSM Parameter Store as it looks cheaper.